presto-main/src/main/java/com/facebook/presto/operator/project/PageProcessor.java

```diff
@@ -15,11 +15,153 @@
 
 import com.facebook.presto.spi.ConnectorSession;
 import com.facebook.presto.spi.Page;
+import com.facebook.presto.spi.block.Block;
+import com.google.common.collect.AbstractIterator;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.Iterators;
 
-public interface PageProcessor
+import javax.annotation.concurrent.NotThreadSafe;
+
+import java.util.List;
+import java.util.Optional;
+
+import static com.facebook.presto.operator.project.PageProcessorOutput.EMPTY_PAGE_PROCESSOR_OUTPUT;
+import static com.google.common.base.Verify.verify;
+import static java.util.Objects.requireNonNull;
+
+@NotThreadSafe
+public class PageProcessor
 {
-    /**
-     * @return filtered and projected page, null if all positions are filtered
-     */
-    PageProcessorOutput process(ConnectorSession session, Page page);
+    static final int MAX_BATCH_SIZE = 8 * 1024;
+    static final int MAX_PAGE_SIZE_IN_BYTES = 4 * 1024 * 1024;
+    static final int MIN_PAGE_SIZE_IN_BYTES = 1024 * 1024;
+
+    private final Optional<PageFilter> filter;
+    private final List<PageProjection> projections;
+
+    private int projectBatchSize = MAX_BATCH_SIZE;
+
+    public PageProcessor(Optional<PageFilter> filter, List<? extends PageProjection> projections)
+    {
+        this.filter = requireNonNull(filter, "filter is null");
+        this.projections = ImmutableList.copyOf(requireNonNull(projections, "projections is null"));
+    }
+
+    public PageProcessorOutput process(ConnectorSession session, Page page)
+    {
+        if (page.getPositionCount() == 0) {
+            return EMPTY_PAGE_PROCESSOR_OUTPUT;
+        }
+
+        if (filter.isPresent()) {
+            SelectedPositions selectedPositions = filter.get().filter(session, filter.get().getInputChannels().getInputChannels(page));
+            if (selectedPositions.isEmpty()) {
+                return EMPTY_PAGE_PROCESSOR_OUTPUT;
+            }
+
+            if (projections.isEmpty()) {
+                return new PageProcessorOutput(page.getRetainedSizeInBytes(), Iterators.singletonIterator(new Page(selectedPositions.size())));
+            }
+
+            if (selectedPositions.size() != page.getPositionCount()) {
+                return new PageProcessorOutput(page.getRetainedSizeInBytes(), new PositionsPageProcessorIterator(session, page, selectedPositions));
+            }
+        }
+
+        return new PageProcessorOutput(
```

Comment:
I am thinking of non-selective filters (e.g: 100 rows out of 8000). This won't be a visible problem now since filtered rows are glued into larger page by page builder. Agree this can be fixed in the future if it becomes a problem. I just wondered if you considered such filter non-selective filters.